docs/library/ussl.rst

```diff
@@ -23,9 +23,29 @@ Functions
    server-side SSL socket should be created from a normal socket returned from
    :meth:`~usocket.socket.accept()` on a non-SSL listening server socket.
 
+   Parameters:
+
+   - ``server_side``: creates a server connection if True, else client connection. A
+     server connection requires a ``keyfile`` and a ``certfile``.
+   - ``cert_reqs``: specifies the level of certificate checking to be performed.
+   - ``ca_certs``: root certificates to use for certificate checking.
+   - ``server_hostname``: specifies the hostname of the server for verification purposes
+     as well for SNI (Server Name Identification).
+   - ``do_handshake``: if True, initiates the TLS handshake and waits for its completion;
+     if False, proceeds without handshake and performs is with the first write making it
+     non-blocking if asyncio is used, see git commit 9c7c082.
+
    Depending on the underlying module implementation in a particular
    `MicroPython port`, some or all keyword arguments above may be not supported.
 
+   ESP32 implementation notes:
+
+   - The esp32 implementation does not support cert_reqs: it never validates certs!
+   - The esp32 implementation supports key-exchange and bidirectional authentication
+     using Pre-Shared Keys. Use KW options ``psk_ident=<identity hint>`` and
```

Comment:
All other parameters of `wrap_socket` are irrelevant for PSK mode, are they? Why not add it as a separate function, then? This way you won't need all these new checks you've added in `wrap_socket` for parameters' correctness.

Reply:
You are correct. However, the reason I added these params is libraries. For example, mqtt_as accepts an `ssl_params` dict and passes that straight through to `wrap_socket`, so the way I'm doing it allows the user to choose PSK without change to mqtt_as. What you describe would require forking mqtt_as to change the call to `wrap_socket` into something different.

Reply:
I see. For your current use of it it's understandable, but I think that libraries should adapt themselves to the core, and not otherwise :)

Also, a simple Pythonic workaround until libraries are updated can be to have a `ssl.py` file containing a `wrap_socket` function that demuxes the call to `ussl.wrap_socket` or `ussl.wrap_socket_psk` based on the arguments.

Reply:
I don't think it's realistic to expect each library to add code to distinguish the form of certificate and to me, a big plus of TLS-PSK is ease of use and that completely vanishes if you have to go fix libraries.

Reply:
Hmm.

> a big plus of TLS-PSK is ease of use

Agreed on this one. Well, at least it should be compile-time configurable, but that's a cleanup to be added later, as you said.